lib/cog/command/service/memory.ex

```diff
@@ -59,100 +61,139 @@ defmodule Cog.Command.Service.Memory do
-                  {:ok, accumulated_value} <- ets_insert(tid, {token, key}, existing_value ++ value),
+    result = with {:ok, existing_value}    <- ets_lookup_join_list(state.memory_table, {token, key}),
+                  {:ok, accumulated_value} <- ets_insert(state.memory_table, {token, key}, existing_value ++ value),
                   do: {:ok, accumulated_value}
 
     {:reply, result, state}
   end
 
-  def handle_call({:replace, token, key, value}, _from, %__MODULE__{tid: tid} = state) do
-    monitor_executor(token, tid)
+  def handle_call({:replace, token, key, value}, _from, state) do
+    monitor_executor(state.monitor_table, token)
 
-    result = ets_insert(tid, {token, key}, value)
+    result = ets_insert(state.memory_table, {token, key}, value)
     {:reply, result, state}
   end
 
-  def handle_call({:delete, token, key}, _from, %__MODULE__{tid: tid} = state) do
-    result = ets_delete(tid, {token, key})
+  def handle_call({:delete, token, key}, _from, state) do
+    result = ets_delete(state.memory_table, {token, key})
     {:reply, result, state}
   end
 
-  def handle_info({:DOWN, monitor_ref, :process, pid, reason}, %__MODULE__{tid: tid} = state) do 
+  def handle_info({:DOWN, _monitor_ref, :process, pid, reason}, state) do
     Logger.debug("Process #{inspect pid} went down (#{inspect reason}); removing its memory storage")
 
-    case :ets.lookup(tid, monitor_ref) do
-      [{^monitor_ref, token}] ->
-        :ets.match_delete(tid, {{token, :_}, :_})
-        :ets.delete(tid, monitor_ref)
-        :ets.delete(tid, token)
-      [] ->
-        Logger.warn("Unknown monitor ref #{inspect monitor_ref} for pid #{inspect pid} going down for #{inspect reason}")
+    case ets_lookup(state.monitor_table, pid) do
+      {:ok, token} ->
+        :ets.match_delete(state.memory_table, {{token, :_}, :_})
+        :ets.delete(state.monitor_table, pid)
+      {:error, :unknown_key} ->
+        Logger.warn("Unknown pid #{inspect pid} was monitored; ignoring")
     end
 
     {:noreply, state}
   end
 
-  defp monitor_executor(token, tid) do
-    case ets_lookup(tid, token) do
-      {:error, :unknown_key} ->
-        case Service.Tokens.process_for_token(token) do
-          {:error, error} ->
-            {:error, error}
-          pid ->
+  def handle_info(:dead_process_cleanup, state) do
+    ets_iterate(state.monitor_table, fn pid, token ->
+      unless Process.alive?(pid) do
+        cleanup_process(state.monitor_table, state.memory_table, pid, token)
+      end
+    end)
+
+    schedule_dead_process_cleanup(@dead_process_cleanup_interval)
+
+    {:noreply, state}
+  end
+
+  defp account_for_existing_processes(monitor_table, memory_table) do
+    ets_iterate(monitor_table, fn pid, token ->
+      case Process.alive?(pid) do
+        true ->
+          Logger.debug("Remonitoring #{inspect pid} for token #{inspect token}")
+          :erlang.monitor(:process, pid)
```

Comment:
I don't think so. The pid is already in the ets table and we're calling this before init returns so all pids in the ets table should be monitored and we can use that to make sure we're not double monitoring.